Entity/Components/Stats/EntityStats.gd

```diff
@@ -0,0 +1,61 @@
+extends Resource
+class_name EntityStats
+## Used as a container for modified runtime stats of an Entity.
+##
+## Use StatComponent to get a reference to the current stats on the Entity.
+## We can increase these stats with buffs, items, or equipment. We shouldn't modify the stats
+## directly, but instead create a modified version of the stats whenever we need them. This is so
+## we don't have to worry about reverting the state of the stats whenever the buff/item/equipment
+## is removed. See get_modified_stats on BuffBase for an example.
+
+## The previous statement will probably change, since we want to modify stats at runtime and we revert the changes when effects are removed.
+
+var modification_count: int = 0
+var offense_modifier_dict: StatDictBase = null
+var defense_modifier_dict: StatDictBase = null
+
+func _init() -> void:
+	offense_modifier_dict = StatDictBase.new()
+	defense_modifier_dict = StatDictBase.new()
+
+# to trigger on the combat end signal, do we have one yet ?
+# do we even have a condition for the end of a combat ?
+func reset_modifier_dict_temp_to_default() -> void:
+	offense_modifier_dict.reset_all_temp_to_default()
+	defense_modifier_dict.reset_all_temp_to_default()
+
+func change_stat(   modifier_dict: StatDictBase, 
+					enum_name: int,
+					new_modification: StatModifiers
+					) -> void:
+	modifier_dict.change_modifier_of_given_name(enum_name, new_modification)
+	modification_count += 1
+
+func _calculate_modified_value_offense(enum_name: int, base_value: int) -> int:
+	var modified_value: int = base_value
+	var modifier: Dictionary = offense_modifier_dict.stat_dict[enum_name].modifiers
+	modified_value += modifier["temporary_add"]
+	modified_value += modifier["permanent_add"]
+	modified_value *= modifier["temporary_multiply"]
+	modified_value *= modifier["permanent_multiply"]
```

Comment:
Does order of operations come into play here?

Reply:
Let's call modified value `m`, then in the order of operation, the modifier stats are `a`, `b`, `c` and `d`
The operation we are doing is `(m + a + b)*c*d = m*c*d + (a+b)*c*d`
If we do the multiplication first, we get: `m*c*d + a + b` which is different. Doing the multiplication after has a stronger effect. The question is, do we want this ?
If I have a buff that says "draw 2 more cards and double the number of card drawn", for 1 card should I draw:
`(1+2)*2 = 6 cards`
or
`1*2 + 2 = 5 cards` ?
What if we have a card that says the above, and another that says " double the number of card drawn and draw 2 more cards" ? That's a different order of operations. Should we always write the cards in the same order (ie + before x or x before +) ?

Reply:
I think either is fine as long as it's consistent. I think we should not have many effects that use multiplication though, because it will make the math confusing for the player. That's more of a design thing though.

Reply:
Yes I also thought it wouldn't be good to have to many effects with multiplications. iirc, StS does times before plus (if you have a weakness effect on an enemy, which reduces damage by 25%, and a relic that reduces one damage for each blow taken, the reduction in percentage by the weakness is applied first)

Reply:
Currently discussing this with the design team